zfsbootmenu/lib/zfsbootmenu-core.sh

```diff
@@ -698,76 +698,71 @@ find_be_kernels() {
-    zdebug "found ${mnt}/boot/${kernel}"
+  # Look for kernels in / and /boot, sorted in version order
+  while read -r kpath; do
+    # Strip mount point from path
+    kpath="${kpath#"${mnt}"}"
+    # Ensure kpath has leading slash
+    kpath="/${kpath#/}"
+    zdebug "found kernel: ${mnt}${kpath}"
+
+    # Extract base name and kernel directory
+    kernel="${kpath##*/}"
+    kdir="${kpath%"${kernel}"}"
+    # Trim trailing slash (note: kdir will be empty if kernel is at root)
+    kdir="${kdir%/}"
+    zdebug "kernel directory: '${kdir}', file: '${kernel}'"
 
     # Kernel "base" extends to first hyphen
     kernel_base="${kernel%%-*}"
+
     # Kernel "version" is everything after base and may be empty
     version="${kernel#"${kernel_base}"}"
     version="${version#-}"
-    zdebug "kernel version: ${version}"
+    zdebug "kernel base: '${kernel_base}', version: '${version}'"
 
     # initramfs images can take many forms, look for a sensible one
     labels=( "$kernel" )
     if [ -n "$version" ]; then
       labels+=( "$version" )
     fi
 
-    local ext pfx lbl i
     # Use a mess of loops instead better brace expansions to control priorities
+    local ext pfx lbl i ipath
     for ext in {.img,""}{"",.{gz,bz2,xz,lzma,lz4,lzo,zstd}}; do
       for pfx in initramfs initrd; do
         for lbl in "${labels[@]}"; do
           for i in "${pfx}-${lbl}${ext}" "${pfx}${ext}-${lbl}"; do
-            if [ -e "${mnt}/boot/${i}" ]; then
-              zdebug "matching ${i} to ${kernel}"
-              echo "${fs} /boot/${kernel} /boot/${i}" >> "${kernel_records}"
-              break 4
-            fi
+            ipath="${kdir}/${i}"
+            [ -e "${mnt}${ipath}" ] || continue
+            zdebug "matching '${i}' to '${kernel}'"
+            echo "${fs} ${kpath} ${ipath}" >> "${kernel_records}"
+            break 4
           done
         done
       done
     done
-  done
+
+  done <<<"$(
+    for k in "${mnt}/boot"/{{vm,}linu{x,z},kernel}{,-*}; do
+      [ -e "${k}" ] && echo "${k}"
+    done | sort -V
+  )"
 
   # No further need for the mount
   umount "${mnt}"
 
-  defaults="$( select_kernel "${fs}" )"
-
-  # shellcheck disable=SC2034
-  IFS=' ' read -r def_fs def_kernel def_initramfs <<<"${defaults}"
+  # Search was successful if at least one kernel can be selected
+  [ -s "${kernel_records}" ] && select_kernel "${fs}" >/dev/null && return 0
 
-  def_kernel_file="${mnt/mnt/default_kernel}"
-
-  # If no default kernel is found, there are no kernels; leave the BE
-  # directory in the same state it would be in had no /boot existed
-  if [ -z "${def_kernel}" ]; then
-    zdebug "no default kernel found for ${fs}"
-    rm -f "${kernel_records}" "${def_kernel_file}"
-    return 1
-  fi
-
-  zdebug "default kernel set to ${def_kernel}"
-  echo "${def_kernel##*/}" > "${def_kernel_file}"
-  return 0
+  # Remove an invalid kernel record if the search failed
+  zdebug "failed to find kernels on ${fs}"
```

Comment:
Should this debug message instead be an error message? Not having any kernels on a filesystem that ZBM expects to find them on is probably worth yelling about.

Reply:
Yes, an error should help leave some breadcrumbs if an expected boot environment is missing kernels. If somebody has a filesystem with `mountpoint=/` that deliberately contains no kernels and unnecessarily triggers the error, `org.zfsbootmenu:active=off` is the fix.

Merged with this change.